[KED-2413] Add a style-guide document (#375)
quantumblacklabs/kedro-viz: STYLE_GUIDE.md

# Front-end style guide

> This document outlines recommended code style, quirks, and best practices for front-end development on Kedro-Viz.

Reviewer: I'd maybe open with something that explains who it's for. Sort of like this perhaps?

```suggestion
Here are our recommendations for anyone contributing front-end development to Kedro-Viz; we cover code style, quirks, and best practices.
```

<!-- @TODO table of contents -->

## Code style and linting

> We use [Prettier](https://prettier.io/), [ESLint](https://eslint.org/), and [stylelint](https://stylelint.io/) to lint our SCSS and JavaScript. For the most part, we use the default config - you can find specific details in the `.prettierrc`, `.eslintrc.json`, and `.stylelintrc` config files. We recommend that you install the Prettier, ESLint, and stylelint plugins for your text-editor, and enable automatic formatting on save.

Reviewer: ```suggestion
We use [Prettier](https://prettier.io/), [ESLint](https://eslint.org/), and [stylelint](https://stylelint.io/) to lint our SCSS and JavaScript. For the most part, we use the default config - you can find specific details in the `.prettierrc`, `.eslintrc.json`, and `.stylelintrc` config files. We recommend that you install the Prettier, ESLint, and stylelint plugins for your text editor, and enable automatic formatting on save.
```

Other than that, we mostly recommend following the [AirBnB JavaScript Style Guide](https://github.com/airbnb/javascript).

##  Browser and device support

> We aim to support recent versions of major modern browsers - i.e. Chome, Firefox, Edge, Safari, and Opera, on both MacOS and Windows. Keeping progressive enhancement in mind, we aim for decent mobile support where possible, i.e. nothing should be broken on mobile, but it's okay for mobile/older browsers to have a slightly degraded experience.

Reviewer: Do you have a preferred way of testing across the range? Maybe this is obvious, or do you use browserstack or something similar? Maybe worth mentioning.

Author: We honestly don't, which is a wider problem and something I should discuss with the QB Labs front-end team. Thanks for noting it!

## Accessibility

> We aim to meet [WCAG 2.1 Level AA](https://www.w3.org/WAI/standards-guidelines/wcag/) standards where possible, while acknowledging that this is a data visualization, so some criteria will be difficult to pass without extraordinary effort. However we will do our best within reason.

Reviewer: ```suggestion
We aim to meet [WCAG 2.1 Level AA](https://www.w3.org/WAI/standards-guidelines/wcag/) standards where possible, while acknowledging that this is a data visualization tool, so some criteria will be difficult to pass without extraordinary effort. 
```

Reviewer: I suggested removing the "However we will do our best within reason." as it's a bit defensive. Is there any particular guidance for tools that are visual, so we can claim to have consulted that?

Author: There is, and that's a good shout. The reason I wrote that is that I know there are other things that we can do that we're not currently doing, as it would be tricky and it's not a high priority at present. So I want to say something to the effect that we will try to make an effort to improve accessibility where reasonable.

Kedro-Viz should be navigable with different input devices (e.g. mouse, keyboard, and touchscreen), with obvious and visible focus states on all interactive elements. Text elements should have sufficient colour contrast and font-size to ensure readability to a minimum Level AA standard. Use [WAI-ARIA](https://www.w3.org/WAI/standards-guidelines/aria/) features where applicable to improve screenreader support.

We recommend checking [axe](https://www.deque.com/axe/) and [Lighthouse](https://developers.google.com/web/tools/lighthouse) to review for potential accessibility errors if unsure.

## Semantic HTML

> Ensure that page content is marked up with [semantic HTML elements](https://html.com/semantic-markup/) in order to ensure that it is as usable and accessible as possible.

Reviewer: ```suggestion
Ensure that page content is marked up with [semantic HTML elements](https://html.com/semantic-markup/) to make it as usable and accessible as possible.
```

Avoid:
```jsx
<div className="form">
  <div>Contact us</div>
  <div>Use this form to send us a message.</div>
  <div>Email address:</div>
  <input type="text" />
  <div onClick={handleSubmit}>Submit</div>
</div>
```
Prefer:
```jsx
<form onSubmit={handleSubmit}>
  <h1>Contact us</h1>
  <p>Use this form to send us a message.</p>
  <label htmlFor="email">Email address:</label>
  <input id="email" type="email" />
  <button>Submit</button>
</form>
```

## CSS

### Class names

> This project uses the [BEM naming convention](http://getbem.com/), to avoid conflicts, reduce rule specificity, and make classes easier to read and understand. [Reducing rule specificity](https://css-tricks.com/strategies-keeping-css-specificity-low/) is particularly important for improving CSS maintainability, so we try to avoid nesting CSS selectors as much as possible.

Reviewer: ```suggestion
This project uses the [BEM naming convention](http://getbem.com/), to avoid conflicts, reduce rule specificity, and make classes easier to read and understand. To impove CSS maintainability we [reduce rule specificity](https://css-tricks.com/strategies-keeping-css-specificity-low/) and avoid nesting CSS selectors as much as possible.
```

Reviewer: I may have broken the meaning of the sentence in my attempt to remove gerunds and simplify. Do you want to make it imperative too, perhaps?

Author: This works! It occurs to me that "avoid nesting CSS selectors as much as possible" is ambiguous: I should probably remove "as much as possible" to keep it clear and concise.

> We use the `pipeline-` prefix for most HTML/CSS classes, to reduce the risk of clashes in the global namespace when this project is imported into other applications. It's important never 

Reviewer: "It's important never" is missing its ending

Author: Thanks! Yeah this part is very much a draft - it's practically scribbled notes. I didn't expect you to get this much into the weeds of reviewing it! 🙈

<!-- @TODO Explain usage of the 'kedro' class, Kedro-UI, and webfonts -->

### Variables

> We use Sass variables to declare colours and reused values wherever useful. It's not necessary to declare every CSS unit value as a variable, but if a value is used in multiple places (for related reasons), then it makes sense to make it a variable - either locally in the component, or application-wide in `src/styles/_variables.scss`. Using a named variable instead of a number can clarify your intent, and help you avoid having to update a value in multiple places later on.

Reviewer: ```suggestion
We use Sass variables to declare colours and reuse values wherever useful. It's not necessary to declare every CSS unit value as a variable, but if a value is used in multiple places (for related reasons), then it makes sense to make it a variable - either locally in the component, or application-wide in `src/styles/_variables.scss`. Use a named variable instead of a number: it can clarify your intent, and help you avoid updates to the value in multiple places later on.
```

Avoid:
```scss
color: #001521;
color: rgba(#001521, 0.5);
color: rgba(0, 0, 0, 0.5);
```

Prefer:
```scss
color: $color-bg-dark-1;
color: rgba($color-bg-dark-1, 0.5);
color: rgba(black, 0.5);
```

### Theme colours

> Kedro-Viz has both light and dark themes, and every component must be styled appropriately for each theme. The themes are set using the same theme classes that Kedro-UI uses - `kui-theme--light` and `kui-theme--dark` - applied on the top-level container element. In order to simplify code and avoid increasing CSS specificity, we use CSS Custom Properties to handle theme colours, and we recommend that you use these where possible.

Reviewer: ```suggestion
Kedro-Viz has both light and dark themes, and every component must be styled appropriately for each theme. The themes are set using the same theme classes that Kedro-UI uses - `kui-theme--light` and `kui-theme--dark` - applied on the top-level container element. To simplify code and avoid increasing CSS specificity, we use CSS Custom Properties to handle theme colours. We recommend that you use these too where possible.
```

Reviewer: Can you mandate it? Should it be more imperative? Also, do you think it's necessary to point to a link about custom properties? Do they also need to be capitalised as a proper noun? So many questions...

Author: You're right: "Please use these wherever possible" is better.

[Custom properties](https://developer.mozilla.org/en-US/docs/Web/CSS/--*) are a core but new-ish part of CSS, so they shouldn't need linking to, but we can. I'd mainly seen them capitalised in headings but you're right, they don't need capitalisation. I'll change that.

Avoid:
```scss
.pipeline-title {
  .kui-theme--light & {
    background: $color-bg-light-1;
  }

  .kui-theme--dark & {
    background: $color-bg-dark-1;
  }
}
```

Prefer:
```scss
.kui-theme--light {
  --color-bg-1: $color-bg-light-1;
}

.kui-theme--dark {
  --color-bg-1: $color-bg-dark-1;
}

.pipeline-title {
  background: var(--color-bg-1);
}
```

<!-- @TODO -->

### Units (em, %, px, etc)

<!-- @TODO -->

## JavaScript

### Imports

> Although it will work in regular development (because this project uses React-Scripts), avoid importing non-standard file-types like `.scss` and `.svg` directly into JavaScript files. This is because these won't work without specific webpack loaders, so when Kedro-Viz is imported into other projects, it will break. The `lib-test` testing suite exists partly for this reason, to check that Kedro-Viz still works when imported as a componnent library into a fairly standard JS app.

Reviewer: ```suggestion
Although it works in regular development (because this project uses React-Scripts), you should avoid importing non-standard file-types like `.scss` and `.svg` directly into JavaScript files. This is because these won't work without specific webpack loaders, and breaks when Kedro-Viz is imported into other projects. The `lib-test` testing suite exists partly for this reason, to check that Kedro-Viz still works when imported as a component library into a fairly standard JS app.
```

###  Testing

In accordance with McKinsey standards, we aim to maintain at least a test coverage of at least 70% averaged across the codebase. As it's not practical to write tests for every JavaScript file, we usually aim for closer to 90% on most files, so that the project average continues to exceed the expected minimum standard.

> We use Jest, Enzyme, and React-Testing-Library for JavaScript testing. Most of the older tests are written with Enzyme, but we are beginning to write more tests using Testing-Library as it is more flexible for testing certain browser APIs. Either is acceptable. To help with mocking the Redux store, Enzyme helper utilities can be found in `/src/utils/state.mock.js`.

Reviewer: Links?

<!-- @TODO Explain the usage of the different datasets in testing, and add more comments explaining how the load-data mock works. -->

## Git and Github

### Commits

> Write commit subjects/titles in the imperative mood. Try to be as specific and detailed as you can, while keeping the subject under 50 characters so that isn't truncated when displayed. If it's a complicated or unclear change and more detail would help future readers understand the commit, please consider adding a detailed description of what you changed and why.

Reviewer: ```suggestion
Write commit subjects/titles in the imperative mood. Try to be as specific and detailed as you can, yet keep the subject under 50 characters so that isn't truncated when displayed. If it's a complicated or unclear change and more detail would help future readers understand the commit, add a detailed description of what you changed and why.
```

We recommend checking [axe](https://www.deque.com/axe/) and [Lighthouse](https://developers.google.com/web/tools/lighthouse) to review for potential accessibility errors if unsure.

## Semantic HTML

Ensure that page content is marked up with [semantic HTML elements](https://html.com/semantic-markup/) to keep it as usable and accessible as possible.

Avoid:

```jsx
<div className="form">
  <div>Contact us</div>
  <div>Use this form to send us a message.</div>
  <div>Email address:</div>
  <input type="text" />
  <div onClick={handleSubmit}>Submit</div>
</div>
```

Prefer:

```jsx
<form onSubmit={handleSubmit}>
  <h1>Contact us</h1>
  <p>Use this form to send us a message.</p>
  <label htmlFor="email">Email address:</label>
  <input id="email" type="email" />
  <button>Submit</button>
</form>
```

## CSS

### Class names

This project uses the [BEM naming convention](http://getbem.com/), to avoid conflicts, reduce rule specificity, and make classes easier to read and understand. To impove CSS maintainability we [reduce rule specificity](https://css-tricks.com/strategies-keeping-css-specificity-low/) and avoid nesting CSS selectors.

> Because Kedro-Viz can be imported into other applications, we avoid ever setting global styles. We use the `pipeline-` prefix for most HTML/CSS classes, to reduce the risk of clashes in the global namespace when this project is imported into other applications. It's important never to use global element selectors or generic classes like `.header`.

Reviewer 2: It's probably a good idea to specifically stress here that whilst CSS files are split up per-component and imported via JS, all CSS is still naturally global in scope. 

CSS modules are pretty common now and it's easy to forget it's not the default (it would be nice though...)

### Units

In general, we try to use `em` for styles that depend on font-size (e.g. text margins), `%` for relative layout positioning, and `px` for absolute layout positioning. Avoid using `rem` or viewport units (e.g. `vh`/`vw`/`vmin`/`vmax`) in code that's included in the package library, as these can cause side-effects if the parent app modifies the base font-size or the app is not full-screen.

### Variables

We use Sass variables to declare colours and reuse values wherever useful. It's not necessary to declare every CSS unit value as a variable, but if a value is used in multiple places (for related reasons), then it makes sense to make it a variable - either locally in the component, or application-wide in `src/styles/_variables.scss`. Using a named variable instead of a number clarifies intent, and helps avoid updates to the value in multiple places later on.

Avoid:

```scss
color: #001521;
color: rgba(#001521, 0.5);
color: rgba(0, 0, 0, 0.5);
```

Prefer:

```scss
color: $color-bg-dark-1;
color: rgba($color-bg-dark-1, 0.5);
color: rgba(black, 0.5);
```

### Theme colours

> Kedro-Viz has both light and dark themes, and every component must be styled appropriately for each theme. The themes are set using the same theme classes that Kedro-UI uses - `kui-theme--light` and `kui-theme--dark` - applied on the top-level container element. To simplify code and avoid increasing CSS specificity, we use CSS custom properties to handle theme colours. We recommend that you use these too where possible - they are defined in the App component.

Reviewer 2: I think it might be a nice idea to leave a blank line after this first sentence to stress the importance of this.

Avoid:

```scss
.pipeline-title {
  .kui-theme--light & {
    background: $color-bg-light-1;
  }

  .kui-theme--dark & {
    background: $color-bg-dark-1;
  }
}
```

Prefer:

```scss
.kui-theme--light {
  --color-bg-1: $color-bg-light-1;
}

.kui-theme--dark {
  --color-bg-1: $color-bg-dark-1;
}

.pipeline-title {
  background: var(--color-bg-1);
}
```

## JavaScript

### Imports

Although it works in regular development (because this project uses React-Scripts), you should avoid importing non-standard file-types like `.scss` and `.svg` directly into JavaScript files. This is because these won't work without specific webpack loaders, and breaks when Kedro-Viz is imported into other projects. The `lib-test` testing suite exists partly for this reason, to check that Kedro-Viz still works when imported as a component library into a fairly standard JS app.

### Testing

In accordance with McKinsey standards, we aim to maintain at least a test coverage of at least 70% averaged across the codebase. As it's not practical to write tests for every JavaScript file, we usually aim for closer to 90% on most files, so that the project average continues to exceed the expected minimum standard.

> We use [Jest](https://jestjs.io/), [Enzyme](https://enzymejs.github.io/enzyme/), and [Testing Library](https://testing-library.com/) for JavaScript testing. Most of the older tests are written with Enzyme, but we are beginning to write more tests using Testing Library as it is more flexible for testing certain browser APIs. Either is acceptable. To help with mocking the Redux store, Enzyme helper utilities can be found in `/src/utils/state.mock.js`.

Reviewer 2: I think some discussion of testing strategy and prioritisation would be worth adding here at some point.

Author: I agree, but can you expand on this? What would you say?
